R-package/tests/testthat/test_lgb.Booster.R

```diff
@@ -1518,3 +1518,73 @@ test_that("LGBM_BoosterGetNumFeature_R returns correct outputs", {
     ncols <- .Call(LGBM_BoosterGetNumFeature_R, model$.__enclos_env__$private$handle)
     expect_equal(ncols, ncol(iris) - 1L)
 })
+
+set.seed(1L)
+data(agaricus.train, package = "lightgbm")
+train <- agaricus.train
+bst <- lightgbm(
+  data = as.matrix(train$data)
+  , label = train$label
+  , params = list(objective = "binary", num_threads = .LGB_MAX_THREADS)
+  , nrounds = 5L
+  , verbose = .LGB_VERBOSITY
+)
+
+# Simplified version of lgb.model.dt.tree()
+get_trees_from_dump <- function(x) {
+  parsed <- jsonlite::fromJSON(
+    txt = x
+    , simplifyVector = TRUE
+    , simplifyDataFrame = FALSE
+    , simplifyMatrix = FALSE
+    , flatten = FALSE
+  )
+  return(lapply(parsed$tree_info, FUN = .single_tree_parse))
+}
+
+get_n_trees <- function(x) {
+  return(length(get_trees_from_dump(lgb.dump(x))))
+}
+
+test_that("num_iteration and start_iteration work for lgb.dump()", {
+  first2 <- get_trees_from_dump(lgb.dump(bst, num_iteration = 2L))
+  last3 <- get_trees_from_dump(
+    lgb.dump(bst, num_iteration = 3L, start_iteration = 2L)
+  )
+  all5 <- get_trees_from_dump(lgb.dump(bst))
+  too_many <- get_trees_from_dump(lgb.dump(bst, num_iteration = 10L))
+
+  expect_equal(c(first2, last3), all5)
+  expect_equal(too_many, all5)
+})
+
+test_that("num_iteration and start_iteration work for lgb.save()", {
+  save_and_load <- function(bst, ...) {
+    model_file <- tempfile(fileext = ".model")
+    lgb.save(bst, model_file, ...)
+    return(lgb.load(model_file))
+  }
+  n_first2 <- get_n_trees(save_and_load(bst, num_iteration = 2L))
+  n_last3 <- get_n_trees(
+    save_and_load(bst, num_iteration = 3L, start_iteration = 2L)
+  )
+  n_all5 <- get_n_trees(save_and_load(bst))
+  n_too_many <- get_n_trees(save_and_load(bst, num_iteration = 10L))
+
+  expect_equal(n_first2, 2L)
+  expect_equal(n_last3, 3L)
+  expect_equal(n_all5, 5L)
+  expect_equal(n_too_many, 5L)
+})
+
+test_that("num_iteration and start_iteration work for save_model_to_string()", {
```

Comment:
Similar to my previous comment, in addition to testing these output sizes, could you please check that all expected `tree_index` values are found in the output?

I think this unit test as currently written would succeed if the `start_iteration` were ignored, since it's only comparing *sizes* and not content.

Reply:
This function just returns a long string. Do we have a parser for it?

Reply:
I'm so sorry, I looked at this tonight and I realize now that this representation doesn't include `tree_index` 😭 

Which I think makes sense.... nothing using this output needs to know which specific iterations from the original training run resulted in the trees in this representation.

So you don't have to check for exact indices. But could you modify the test to check for a specific number of trees?

For example, `save_model_to_string(num_iteration=2L)` should have exactly 2 trees in it (or `2 * num_classes` trees for multiclass classification).

You can search for the number of lines matching the regular expression `Tree=[0-9]+`.

The output contains sections like the following, 1 per tree:

```text
...

Tree=0
num_leaves=4
num_cat=0
split_feature=40 26 57
split_gain=634807 418894 240563
threshold=-0.11196126005669692 -0.24042318053852677 0.27519829716972061
...

Tree=1
num_leaves=4
num_cat=0
split_feature=71 74 26
split_gain=518846 441479 136866
threshold=0.27391479370334465 0.013066561649385923 -0.40442439996877116
...

Tree=2
num_leaves=4
num_cat=0
split_feature=40 26 71
split_gain=477521 306133 206409
threshold=-0.11196126005669692 -0.24042318053852677 0.084521718692538536
...
```

<details><summary>Python code to generate that (click me)</summary>

```python
import lightgbm as lgb
from sklearn.datasets import make_regression

X, y = make_regression()
bst = lgb.train(
    params={
        "objective": "regression",
        "num_iterations": 10
    },
    train_set=lgb.Dataset(X, y, params={"min_data_in_bin": 1})
)
print(bst.model_to_string(start_iteration=3, num_iteration=4))
```

</summary>

Reply:
Good idea! Actually, the corresponding unit test fails... "last3" tree indices are expected to be 2:4, while the function returns 0:3.

Reply:
Right, this is what I meant by this statement:

> you don't have to check for exact indices.

So the first tree in the output of `save_model_to_string()` will always have `Tree=0`, the second will always have `Tree=1`, etc.

For this test, instead of checking for exact indices like this:

```r
expect_equal(extract_indices_from_string(last3), 2L:4L)
```

I'm asking you to do something for `save_model_to_string()` identical to the tests you have above for `lgb.save()`... just checking that the **count** of trees is exactly what's expected. That's a stronger test than what was here previously, which was just checking that asking for 3 iterations generated more *characters* than asking for 2 trees.

Reply:
Got it, thanks for iterating! 

(Originally, I was not aware that `save_model_to_string()` behaves like `save_model()` regarding tree indices.)